chocolatey/choco
# (maint) Implement backup and restore strategy to fix locked files

## Description Of Changes

This pull request adds/updates the backup and restore strategy that was in place to how it was determined internally it should work.
Additional changes are made to ensure that this strategy is used during install (only forced), upgrades and uninstalls as well.

## Motivation and Context

This backup and restore strategy is done in part to have it follow what we expect to happen, to refer to what happens when a file is locked
and to ensure as best as we can that installed packages are lost if an upgrade or uninstall fails.

## Testing

I will assume you have internal choco-test-packages repository enabled, and the feature `allowGlobalConfirmation` is disabled.

Install tests:

1. Run `choco install failingdependency`.
2. Once asked to run install script, verify no backup has been created and there is no package in the lib-bad directory.
3. Run the uninstall script and wait for it to fail.
4. Verify the package is not located in `lib` but is available in `lib-bad` with all of its files.
5. Run `choco install failingdependency -n`
6. Create a new file inside of `lib\failingdependency`
7. `Run `choco install failingdependency --force`
8. Once asked to run install script, verify there is a backup created and it contains your custom faile, there is no package in the lib-bad directory and in the `lib` directyr your custom file is not available.
9. Run the uninstall script and wait for it to fail.
10. Verify the package is located in the `lib` directory with your custom file, there is no backup directory available anymore and the package is located in `lib-bad` without your custom file.

Upgrade tests

TODO, see E2E tests in the meantime

Uninstall tests

TODO, see E2E tests in the meantime (in progress)

### Operating Systems Testing

- Windows 10
- Windows Server 2019

## Change Types Made


* [x] Documentation changes. (internally only)

## Change Checklist

* [x] Requires a change to the documentation. (internally)
* [x] Documentation has been updated. (internally)
* [ ] All new and existing tests passed? (not all E2E tests have been ran)

## Related Issue

- Fixes #2690 (due to the backup strategy has changed)
- https://app.clickup.com/t/20540031/PROJ-529
